sklearn/metrics/plot/roc_curve.py

```diff
@@ -0,0 +1,181 @@
+from .. import auc
+from .. import roc_curve
+
+from ...utils import check_matplotlib_support  # noqa
+
+
+class RocCurveVisualizer:
+    """ROC Curve visualization
+
+    Parameters
+    ----------
+    estimator : estimator instance
+        Trained classifier.
+
+    X : {array-like, sparse matrix}, shape (n_samples, n_features)
+        Input values.
+
+    y : array-like, shape (n_samples, )
+        Target values.
+
+    pos_label : int str or None, default=None
+        The label of the positive class.
+        When `pos_label=None`, if y_true is in {-1, 1} or {0, 1},
+        `pos_label` is set to 1, otherwise an error will be raised.
+
+    sample_weight : array-like of shape = [n_samples], optional
+        Sample weights.
+
+    drop_intermediate : boolean, default=True
+        Whether to drop some suboptimal thresholds which would not appear
+        on a plotted ROC curve. This is useful in order to create lighter
+        ROC curves.
+
+    response_method : 'predict_proba', 'decision_function', or 'auto' \
+    default='auto'
+        Method to call estimator to get target scores
+
+    Attributes
+    ----------
+    fpr_ : ndarray
+        False positive rate.
+    tpr_ : ndarray
+        True positive rate.
+    auc_ :
+        Area under ROC curve.
+    estimator_name_ : str
+        Name of estimator.
+    line_ : matplotlib Artist
+        ROC Curve.
+    ax_ : matplotlib Axes
+        Axes with ROC curv
+    figure_ : matplotlib Figure
+        Figure containing the curve
+    """
+
+    def __init__(self, estimator, X, y, pos_label=None, sample_weight=None,
+                 drop_intermediate=True, response_method="auto"):
+        """Computes and stores values needed for visualization"""
+
+        if response_method != "auto":
+            prediction_method = getattr(estimator, response_method, None)
+            if prediction_method is None:
+                raise ValueError(
+                    "response method {} not defined".format(response_method))
+        else:
+            predict_proba = getattr(estimator, 'predict_proba', None)
+            decision_function = getattr(estimator, 'decision_function', None)
+            prediction_method = predict_proba or decision_function
+
+        if prediction_method is None:
+            if response_method == 'predict_proba':
+                raise ValueError('The estimator has no predict_proba method')
+            else:
+                raise ValueError(
+                    'The estimator has no decision_function method')
+
+        y_pred = prediction_method(X)
+
```

Comment:
I wonder if we should move the computing part outside of the init, in a `compute` (fit ?) method which looks more like sklearn. Then `plot_XXX` would do 
```
viz = XXXVisualizer()
viz.compute()
viz.plot()
```

Reply:
I have considered this and sided with doing the computation in `__init__` to simplify the `Visualizer` API. 

With your proposal the API looks like:

```py
viz = ROCVisualizer(response_method="predict_proba")
viz.compute(estimator, X, y)
viz.plot()
```

This comes down how useful a `Visualizer` object without computing. Our estimators make use of this "delay" computation through `fit` because we want to put them into `Pipelines`, etc. For `Visualizer`, I am not envisioning a need for delaying computation.

Reply:
I agree with @thomasjpfan. I don't see a point in building visualization pipelines as yellowbrick does. The only purpose seems to be to "look more like sklearn", resulting in a harder to understand api and more complex code.
I don't think "looking like an estimator" is a good goal if you do something entirely different.